src/systems/particle_emitter2/ParticleEmitter2.cc

```diff
@@ -160,71 +142,71 @@ void ParticleEmitter2::PreUpdate(const ignition::gazebo::UpdateInfo &_info,
             {
-              topic = data.value(0);
+              if (data.key() == "topic" && !data.value().empty())
+              {
+                topic = data.value(0);
+              }
             }
           }
-        }
 
-        // If a topic has not been specified, then generate topic based
-        // on the scoped name.
-        topic = !topic.empty() ? topic :
-          topicFromScopedName(_entity, _ecm) + "/cmd";
+          // If a topic has not been specified, then generate topic based
+          // on the scoped name.
+          topic = !topic.empty() ? topic :
+            topicFromScopedName(_entity, _ecm) + "/cmd";
 
-        // Subscribe to the topic that receives particle emitter commands.
-        if (!this->dataPtr->node.Subscribe(
-              topic, &ParticleEmitter2Private::OnCmd, this->dataPtr.get()))
-        {
-          ignerr << "Error subscribing to topic [" << topic << "]. "
-            << "Particle emitter will not receive updates." << std::endl;
-          return false;
-        }
-        ignmsg << "Particle emitter["
-          << scopedName(_entity, _ecm, "::", false) << "] subscribed "
-          << "to command messages on topic[" << topic << "]\n";
-
-        // Store the topic name so that we can apply user commands
-        // correctly.
-        this->dataPtr->emitterTopicMap[topic] = _entity;
-        return true;
-      });
-
-  // Populate teh service message on demand
-  if (this->dataPtr->serviceRequest)
-  {
-    std::unique_lock<std::mutex> lockCv(this->dataPtr->serviceMutex);
-    this->dataPtr->serviceMsg.Clear();
+          // Subscribe to the topic that receives particle emitter commands.
+          if (!this->dataPtr->node.Subscribe(
+                topic, &ParticleEmitter2Private::OnCmd, this->dataPtr.get()))
+          {
+            ignerr << "Error subscribing to topic [" << topic << "]. "
+              << "Particle emitter will not receive updates." << std::endl;
+            return false;
+          }
+          ignmsg << "Particle emitter["
+            << scopedName(_entity, _ecm, "::", false) << "] subscribed "
+            << "to command messages on topic[" << topic << "]\n";
 
-    // Get all the particle emitters
-    _ecm.Each<components::ParticleEmitter>([&](const Entity & /*_entity*/,
-          const components::ParticleEmitter *_emitter)->bool
-        {
+          // Store the topic name so that we can apply user commands
+          // correctly.
+          this->dataPtr->emitterTopicMap[topic] = _entity;
+
+          // Store the emitter information in the service message, which
+          // can then be used in the particle_emitters service.
           msgs::ParticleEmitter *emitterMsg =
             this->dataPtr->serviceMsg.add_particle_emitter();
           emitterMsg->CopyFrom(_emitter->Data());
+          math::Pose3d pose = worldPose(_entity, _ecm);
+          msgs::Set(emitterMsg->mutable_pose(), pose);
```

Comment:
initially I thought since particle emitter is now s on the same level as sensor, visual and collision, the pose field would be in parent frame so that they are consistent. Is it in world frame here because the particle emitter is missing a parent id field (and so the subscribers can't compute the pose themselves in world frame)?

Reply:
Using relative pose and added frame information in d57cfa204f4f713031e844edda0a6f003940f60b